### PR TITLE
Compat with typed pyproject-hooks

### PR DESCRIPTION
**Pull request by layday**

### CODE REVIEW

File: src/build/__init__.py [hunk 1]
```diff
@@ -40,7 +40,18 @@
...
+class RunnerType(Protocol):
+    """A protocol for the subprocess runner."""
+
+    def __call__(
+        self,
+        cmd: Sequence[str],
+        cwd: str | None = None,
+        extra_environ: Mapping[str, str] | None = None,
+    ) -> None:
+        ...
```
Comment (henryiii) [inline src/build/__init__.py:52]: We can set `...` as uncoverable in coverage's config.

File: src/build/_util.py [hunk 2]
```diff
@@ -1,10 +1,19 @@
...
+import abc
 import re
 import sys
+import typing
 
 from collections.abc import Iterator, Set
 
+if sys.version_info >= (3, 8):
+    from typing import Protocol as Protocol
+elif typing.TYPE_CHECKING:
```
Comment (henryiii) [inline src/build/_util.py:12]: This too

**Comment by pradyunsg** [comment]
Consider using `from pyproject_hooks import SubprocessRunner` behind a `TYPE_CHECKING`  conditional instead of redefining the protocol.

**Comment by layday** [comment]
> Consider using `from pyproject_hooks import SubprocessRunner` behind a `TYPE_CHECKING` conditional instead of redefining the protocol.

`SubprocessRunner` is not public :)

**Comment by layday** [comment]
All of our other types are also available at runtime, so it'd be odd for this one not to be.

**Comment by pradyunsg** [comment]
Look at the PR again. ;)

**Comment by gaborbernat** [comment]
Seems it stalled, @layday let me know if you want to pick it up again and we can reopen.

**Comment by layday** [comment]
This was in draft until a version of pyproject-hooks is released with type hints, which hasn't happened yet.

**Pull request closed without merging**